RHDEVDOCS 4233 add information on Pipelines authentication using workspaces, container registries

mramendi opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s):
<!--- Specify the version or versions of OpenShift your PR applies to. -->
cp to pipelines-docs-1.10, pipelines-docs-1.11, pipelines-docs-1.12, pipelines-docs-1.13, pipelines-docs-1.14

Issue:
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

RHDEVDOCS 4233

Link to docs preview:
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->
https://70384--docspreview.netlify.app/openshift-pipelines/latest/secure/authenticating-pipelines-repos-using-secrets

QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

vdemeester commented on modules/op-using-ssh-authentication-in-git-type-tasks.adoc:
```diff
@@ -1,11 +1,30 @@
+    - name: example-git-step
+#     ...
+      script:
+        ln -s tekton/home/.ssh /root/.ssh
```
`ln -s /tekton/home/.ssh …` or better `ln -s $HOME/.ssh …`.

vdemeester reviewed (commented):
LGTM, one nit.

vdemeester reviewed (commented):
I also agree that the `modules/op-limiting-secret-access-to-specific-steps.adoc` could be removed.

vdemeester commented:
/approve

mramendi commented:
/label peer-review-needed

skopacz1 commented:
/label peer-review-in-progress

skopacz1 commented on modules/op-using-secrets-as-a-nonroot-user.adoc:
```diff
@@ -15,4 +15,4 @@ In such scenarios, consider the following aspects of executing task runs and pip
 * SSH authentication ignores the `$HOME` environment variable. So you must or symlink the appropriate secret files from the `$HOME` directory defined by {pipelines-shortname} (`/tekton/home`), to the non-root user's valid home directory.
 
-In addition, to configure SSH authentication in a non-root security context, refer to the link:https://github.com/tektoncd/pipeline/blob/main/examples/v1/taskruns/authenticating-git-commands.yaml[example for authenticating git commands].
+In addition, to configure SSH authentication in a non-root security context, refer to the `git-clone-and-check` step in the link:https://github.com/tektoncd/pipeline/blob/main/examples/v1/taskruns/authenticating-git-commands.yaml[example for authenticating git commands].
```
I know that this already existed when you made this change, but per our [contribution guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#links-to-external-websites) we "cannot link to a repository that is hosted on www.github.com". "pipeline" is a repository for the tektoncd organization, right? Do you happen to know if this link was included by some special exception or not? If not, I feel like this should ideally be removed, as unfortunate as it would be to get rid of this seemingly useful example.

mramendi replied:
I was under the impression that repositories maintained by Red Hat engineers were excepted, will await @adellape comment

adellape replied:
My understanding was https://github.com/redhat-developer was the only exception.

But yes, otherwise this reference should either be removed or ported and QE-verified to this repo in some way.

mramendi replied:
I have requested Pipelines QE feedback on these examples.

mramendi replied:
This one issue kept the PR dormant for two months. But it is now resolved, I have the new repo.

skopacz1 commented on modules/op-using-ssh-authentication-in-git-type-tasks.adoc:
```diff
@@ -1,11 +1,30 @@
+= Using Git SSH authentication in tasks
 
-When invoking Git commands, you can use SSH authentication directly in the steps of a task. SSH authentication ignores the `$HOME` variable and only uses the user's home directory specified in the `/etc/passwd` file. So each step in a task must symlink the `/tekton/home/.ssh` directory to the home directory of the associated user.
+You can invoke directly Git commands in the steps of a task and use SSH authentication, but an additional step is required.
```
Tiny adverb position and [active voice](https://www.ibm.com/docs/en/ibm-style?topic=grammar-verbs#voice) nit:
```suggestion
You can directly invoke Git commands in the steps of a task and use SSH authentication, but you must perform an additional step.
```

skopacz1 commented on modules/op-using-ssh-authentication-in-git-type-tasks.adoc:
```diff
@@ -1,11 +1,30 @@
+
+However, explicit symlinks are not necessary when you use a pipeline resource of the `git` type or the `git-clone` task available in the Tekton catalog.
 
 As an example of using SSH authentication in `git` type tasks, refer to link:https://github.com/tektoncd/pipeline/blob/main/examples/v1/taskruns/authenticating-git-commands.yaml[authenticating-git-commands.yaml].
```
This is even further out of scope for this review, so I'm only pointing it out this time, but this is another instance of a problematic link to a GitHub-hosted repository.

mramendi replied:
Again, I believe the repo is maintained by Red Hat engineers. There are many links to this and other similar repos in this doc set and removing them would be a major problem. As far as I am aware, the intend of the guidance is to disallow repos that we don't control and that therefore can be volatile.

adellape replied:
The examples could still change out from under us and not necessarily have been QE verified, particularly for the docs version they might be appearing in (especially if just targeting the `main` branch like this example is).

mramendi replied:
I have requested Pipelines QE feedback on these examples.

skopacz1 commented on secure/authenticating-pipelines-repos-using-secrets.adoc:
```diff
@@ -0,0 +1,50 @@
+You can use service accounts to provide secrets for authentication with Git repositories and container repositories.
+
+You can associate a secret with a service account. The information in the secret becomes available to the tasks that run under this service account.
+
```
I think since all the procedural modules in this section use the oc CLI, it would be good to have a Prereqs section defining that as a prereq:
```suggestion

[id="providing-secrets-using-service-accounts-prereqs_{context}"]
=== Prerequisites
* You have installed the OpenShift CLI (`oc`)

```

Not honestly sure if the id/anchor is needed. Optionally, if you prefer then you can add this as a prereq in each procedural module that includes the use of `oc` instead of putting it here.

mramendi replied:
Did this before the subdivisions.

skopacz1 commented on modules/op-configuring-registry-authentication-sa.adoc:
```diff
@@ -0,0 +1,83 @@
+For a pipeline to retrieve container images from a registry or push container images to a registry, you must configure the authentication for that registry.
+
+To configure registry authentication for a pipeline, create an authentication secret with the Docker configuration file, associate this secret with a service account, and associate this service account with a `TaskRun` or `PipelineRun` resource.
+
```
I'd add a prereq here for the oc CLI:
```suggestion

.Prerequisites
* You have installed the OpenShift CLI (`oc`)

```

skopacz1 reviewed (commented):
Left comments mostly related to style guide and formatting nuances, but this is seriously nice work, great job! Let me know if you have any questions about my feedback.

I am still (relatively) new to reviewing, so please wait until @adellape approves of my review before implementing any feedback. Thanks!

mramendi commented on modules/op-configuring-container-registry-authentication-using-workspaces.adoc:
```diff
@@ -0,0 +1,73 @@
+// This module is included in the following assemblies:
+// * secure/authenticating-pipelines-and-tasks-using-secrets.adoc
+
+[id="op-configuring-container-registry-authentication-using-workspaces_{context}"]
```
I was under the impression that this was no longer a requirement but I can be mistaken, will await @adellape comment

mramendi commented on modules/op-configuring-basic-authentication-for-git.adoc:
```diff
@@ -51,9 +55,9 @@ secrets:
+. Create the YAML manifest for the task run or pipeline run in the `run.yaml` file, associate the service account with the task run or pipeline run. Use one of the following examples:
 +
-* Associate the service account with a task run:
+* Associate the service account with a `TaskRun` resource:
```
I disagree regarding numbered steps, I think at last one ISG example has unnumbered ones - the one with "Click Configure to do any of the following tasks"

mramendi replied:
I also think that using the imperative would not be consistent with "Use one of the following examples:" as the introduction (which you also suggested and I do agree). However there was no second-level offset, I fixed that now.

mramendi commented on _topic_maps/_topic_map.yml:
```diff
@@ -106,8 +106,8 @@ Topics:
-- Name: Authenticating pipelines using git secret
-  File: authenticating-pipelines-using-git-secret
+- Name: Authenticating pipelines with repositories using secrets
+  File: authenticating-pipelines-repos-using-secrets
```
I considered this but I kinda think the addition of containers is too important

mramendi commented on modules/op-configuring-container-registry-authentication-using-workspaces.adoc:
```diff
@@ -0,0 +1,73 @@
+
+. In the steps of the task, access the directory using the path in the `$(workspaces.<workspace_name>.path)` environment variable, for example, `$(workspaces.dockerconfig.path)`
+
+. When running the task, specify the secret for the named workspace, as in the following example command:
++
```
I made a shot at this suggestion. Not sure if this works better, though.

mramendi commented on modules/op-configuring-container-registry-authentication-using-workspaces.adoc:
```diff
@@ -0,0 +1,73 @@
+[source, terminal]
+----
+$ oc create secret generic my-registry-credentials \ # <1>
+  --from-file=config.json=/path/to/config.json # <2>
```
Your suggestion led to inconsistency with the precious line and callout, so I made it fully consistent instead

mramendi commented on modules/op-configuring-git-ssh-authentication-using-workspaces.adoc:
```diff
@@ -0,0 +1,113 @@
+[source, terminal]
+----
+$ oc create secret generic my-github-ssh-credentials \ # <1>
+  --from-file=id_ed25519=/path/to/.ssh/id_ed25519 \ # <2>
+  --from-file=known_hosts=/path/to/.ssh/known_hosts # <3>
```
Again. made all lines fully consistent instead

mramendi commented on modules/op-types-annotation-secrets.adoc:
```diff
@@ -0,0 +1,114 @@
+
+You can create `kubernetes.io/dockercfg` and `kubernetes.io/dockerconfigjson` secrets from an existing configuration file, as in the following example:
+
+.Example: Command for creating a secret for authenticating to a container repository from an existing configuration file
```
Unfortunately this change would make it inconsistent with previous examples in the file

mramendi commented:
@skopacz1 I made the changes I was reasonably sure I had an idea about and will now await @adellape review.

Note: I am aware of the need to squash the commits before requesting the merge review. I did not yet squash the commits because there might be more change requests in this peer review.

adellape reviewed (commented):
Added a few more comments and replies, but peer review LGTM 👍

mramendi commented:
@adellape thanks! I have made the requested changes. I have asked our QE regarding the tektoncd github repo.

aireilly commented:
This branch has been repointed to target pipelines-docs-main. pipelines-docs is no longer in use.

mramendi commented:
This was dormant for two months after a completed peer review, because the peer review pointed out a link to an upstream repository and getting an RH repository to replace it took time. This is done now. Other peer review notes were fixed back then, and there were no further changes. Requesting merge.

sheriff-rh commented on secure/authenticating-pipelines-repos-using-secrets.adoc:
```diff
@@ -0,0 +1,55 @@
+
+include::modules/op-configuring-registry-authentication-sa.adoc[leveloffset=+2]
+
+=== Additional considerations for authentication using service accounts
```
No anchor ID on this H3?

sheriff-rh commented on about/understanding-openshift-pipelines.adoc:
```diff
@@ -35,6 +35,13 @@ include::modules/op-about-taskrun.adoc[leveloffset=+2]
 include::modules/op-about-pipelinerun.adoc[leveloffset=+2]
+
+[role="_additional-resources"]
+.Additional resources
```
Needs an anchor ID. https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies

sheriff-rh commented on create/creating-applications-with-cicd-pipelines.adoc:
```diff
@@ -55,6 +55,12 @@ include::modules/op-mirroring-images-to-run-pipelines-in-restricted-environment.
 include::modules/op-running-a-pipeline.adoc[leveloffset=+1]
 
+[role="_additional-resources"]
+.Additional resources
```
Needs an anchor ID. https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies

sheriff-rh reviewed (changes requested):
I noticed a few missing anchor IDs, which may not seem like a big deal but this is the kind of thing that if it doesn't get addressed now, will become a problem on the customer portal publication - even if it passes a github test build.

mramendi commented:
@sheriff-rh I have added the anchor IDs. For the additional resources, I added them to every additional resources section in the affected assemblies in order to ensure consistency.

mramendi commented:
/label merge-review-needed

mramendi commented:
I have made the requested change but am unable to reach @sheriff-rh , I hope someone can merge this

sheriff-rh reviewed (approved):
Changes LGTM, merging.

sheriff-rh commented:
/cherrypick pipelines-docs-1.14

sheriff-rh commented:
/cherrypick pipelines-docs-1.13

sheriff-rh commented:
/cherrypick pipelines-docs-1.12

sheriff-rh commented:
/cherrypick pipelines-docs-1.10

sheriff-rh commented:
Looks like a merge conflict on 1.10 @mramendi - I was able to merge 1.11-1.14 cleanly though :+1:

mramendi commented:
Thanks! Let's forget 1.10

mramendi commented:
Pipelines 1.10 is no longer supported, so I will not adapt the change to Pipelines 1.10 docs, pending their possible removal or marking as unsupported.
